### PR TITLE
Fix issue 90

### PR DESCRIPTION
**Pull request by tariqkurd-repo**
Changes are:

1. pcc is no longer accessible as a CSR
2. the pcc value is UNSPECIFIED in debug mode, and ASR permission is implicitly granted
3. add dinfc CSR as a debug mode only source of infinity
4. remove ASR permission from all debug mode CSRs as it's implicitly granted
5. remove "M-mode" as a prerequisite for (e.g.) MTVECC as M-mode must always be present

### CODE REVIEW

File: src/insns/pcrel_debug_warning.adoc [hunk 1]
```diff
@@ -1,2 +1,4 @@
 NOTE: The instructions on this page are either PC relative or may update the
-<<pcc>>. Therefore an implementation may make them illegal in debug mode.
+<<pcc>>. Therefore an implementation may make them illegal in debug mode. If
+not then the value of the <<pcc>> in debug mode is UNSPECIFIED by this
```
Comment (arichardson) [inline src/insns/pcrel_debug_warning.adoc:3]:
```suggestion
they are supported then the value of the <<pcc>> in debug mode is UNSPECIFIED by this
```

Comment (arichardson) [reply src/insns/pcrel_debug_warning.adoc:3]: I will need to read the debug spec to ensure that whatever we say here matches the debug spec, but maybe we say PCC has almighty permissions/bounds and whatever address mandated by the debug spec?

File: src/debug-integration.adoc [hunk 2]
```diff
@@ -107,3 +106,13 @@ The <<dscratch1c>> register is a CLEN-bit plus tag bit extension to
 
 .Debug scratch 1 capability register
 include::img/dscratch1creg.edn[]
+
+[#dinfc,reftext="dinfc"]
+==== Debug Infinite Capability Register (dinfc)
+
+The <<dinfc>> register is a CLEN-bit plus tag bit CSR only accessible in debug mode.
+It returns the value of the <<infinite-cap>> capability. It is read/write but with no
```
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:114]: probably not - and there are many other similar CSRs which are read/write with no writeable fields. The point is regularity - all other Debug mode CSRs are R/W - and having this one cause an exception on writing is irregular (I've already been moaned at for adding custom CSRs in an irregular way - so it's much easier to keep the whole thing consistent).

[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:114]: @Timmmm you complained about irregularity in CSR permissions before  - what's your opinion?

File: src/debug-integration.adoc [hunk 3]
```diff
@@ -50,15 +50,14 @@ Upon entry to debug mode, cite:[riscv-debug-spec], does not specify how to
 update the PC, and says PC relative instructions may be illegal. This concept
 is extended to include any instruction which updates <<pcc>>.
 
+As a result, the value of <<pcc>> is UNSPECIFIED in debug mode, and has no
+architectural effect if PC relative instructions are illegal. Therefore
+<<asr_perm>> is implicitly granted for access to all CSRs in debug mode.
```
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:55]: >`but that does not mean it is unspecified`

that's exactly what it means, they haven't specified it, but also they haven't written that they haven't specified it they just haven't bothered to mention it.

In contrast, we are saying that we haven't specified it, and admitting to that. That's the only difference.

File: src/debug-integration.adoc [hunk 4]
```diff
@@ -50,15 +50,14 @@ Upon entry to debug mode, cite:[riscv-debug-spec], does not specify how to
...
+
 <<dpcc>> (and consequently <<dpc>>) are updated with the
 capability in <<pcc>> whose address field is set to the address of the next
-instruction to be executed as described in cite:[riscv-debug-spec].
-
-Additionally, the <<pcc>> is updated as follows:
-
-* All metadata is set to the <<infinite-cap>> capability
-  ** The <<pcc>> may be used as a source of the <<infinite-cap>> capability in
-debug mode to allow other capabilities to be created and written into memory.
```
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:61]: no because the pc is `unspecified`

File: src/debug-integration.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (Timmmm) [reply src/debug-integration.adoc:114]: I think if you look at this table then all debug-mode-only registers *have* to be read/write, so I think this is the only correct option.

image

File: src/csv/CHERI_CSR.csv [hunk 5]
```diff
@@ -25,9 +25,8 @@ direct write if address didn't change","✔","","✔","","✔","✔","S-mode","S
...
+"mtdc","0x74c","","","M","MRW, <<asr_perm>>","tag=0, otherwise undefined","","","","","","","✔","","None","Machine Trap Data Capability (scratch register)","","","","","","","","","","","","","","","","","","","","",""
 "stdc","0x163","","","S","SRW, <<asr_perm>>","tag=0, otherwise undefined","","","","","","","✔","","S-mode","Supervisor Trap Data Capability (scratch register)","","","","","","","","","","","","","","","","","","","","",""
 "ddc","0x416","","","U","URW","<<infinite-cap>>","","","","✔","","","✔","","none","User Default Data Capability","","","","","","","","","","","","","","","","","","","","",""
-"pcc","0xcb0","","","U","URO","<<infinite-cap>>
-(address = boot address)","","","✔","","","","✔","✔","none","User Program Counter Capability (to allow reading in legacy mode)","","","","","","","","","","","","","","","","","","","","",""
+"dinfc","0x7bc","","","D","DRW","<<infinite-cap>>","","","","","","✔","✔","✔","Sdext","Source of <<infinite-cap>> capability in debug mode, writes are ignored","","","","","","","","","","","","","","","","","","","","",""
```
Comment (Timmmm) [inline src/csv/CHERI_CSR.csv:32]: Is this meant to be 0x7be? Looks like `dddc` already uses 0x7bc?

File: src/debug-integration.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc]: spec updated, please have a look

File: src/debug-integration.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:61]: spec updated please have a look

File: src/csv/CHERI_CSR.csv (hunk 5 above)
Comment (tariqkurd-repo) [reply src/csv/CHERI_CSR.csv]: it was supposed to be 0x7bd, fixed now

File: src/debug-integration.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tariqkurd-repo) [reply src/debug-integration.adoc]: I've made it R/W for future expansion now, so that if multiple roots are necessary then this can cover it

File: src/csv/CHERI_CSR.csv (hunk 5 above)
Comment (Timmmm) [reply src/csv/CHERI_CSR.csv]: Oh yeah I forgot the alphabet :-D

File: src/debug-integration.adoc [hunk 6]
```diff
@@ -107,3 +110,22 @@ The <<dscratch1c>> register is a CLEN-bit plus tag bit extension to
...
+[#dinfc,reftext="dinfc"]
+==== Debug Infinite Capability Register (dinfc)
+
+The <<dinfc>> register is a CLEN-bit plus tag bit CSR only accessible in debug mode.
+
+The reset value is the <<infinite-cap>> capability. The M field (if
+{cheri_mode_ext_name} is implemented) represents the current mode and
+will be only updated if <<MODESW>> is executed.
+
+<<dinfc>> is read/write but with no writeable fields, and so writes are
```
Comment (arichardson) [inline src/debug-integration.adoc:123]: Wouldn't it be simpler to declare it is read-only and if needed relaxing that restriction in the future? I'm not sure I see the need for writing even with multiple formats.

File: src/debug-integration.adoc [hunk 7]
```diff
@@ -107,3 +110,22 @@ The <<dscratch1c>> register is a CLEN-bit plus tag bit extension to
...
+ignored.
+
+NOTE: A future version of this specification may add writeable fields to allow creation
```
Comment (arichardson) [inline src/debug-integration.adoc:126]: If there are multiple roots in the future couldn't we just add new debug-"infinite code cap, infinite data cap" CSRs and have this one be one of them? Writing to this CSR seems a bit odd since it's then stateful and we have to keep track of that state while in debug mode?

**Review by arichardson — COMMENTED** [review]
Not sure about the dinfc R/W designation otherwise LGTM.

File: src/debug-integration.adoc (hunk 6 above)
Comment (Timmmm) [reply src/debug-integration.adoc:142]: It can't be read-only because all the debug-only CSRs are also in the read/write CSRs section.

Does make me wonder why they bothered with the RW/RO distinction at all tbh.

File: src/debug-integration.adoc (hunk 7 above)
Comment (Timmmm) [reply src/debug-integration.adoc:145]: The issue is there's only space for 16 debug-only CSRs and with this change 14 of them are used.

File: src/debug-integration.adoc [hunk 8]
```diff
@@ -48,17 +67,20 @@ include::img/dpccreg.edn[]
 
 Upon entry to debug mode, cite:[riscv-debug-spec], does not specify how to
 update the PC, and says PC relative instructions may be illegal. This concept
-is extended to include any instruction which updates <<pcc>>.
+is extended to include any instruction which updates <<pcc>>, which refers to
```
Comment (sorear) [inline src/debug-integration.adoc:70]:
```suggestion
is extended to include any instruction which reads or updates <<pcc>>, which refers to
```

AUIPC doesn't update pcc

File: src/debug-integration.adoc (hunk 6 above)
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:142]: I think it's more problematic to change teh type of the register and I'm not aware of that ever happening. The difference is trivial really - either you return an error to the debug host on write or you ignore the write. Also the RISC-V debug arch. doesn't include _any_ read only debug mode CSRs, so having it as DRW fits much better without being irregular.

File: src/debug-integration.adoc (hunk 7 above)
Comment (sorear) [reply src/debug-integration.adoc:145]: dcsr, dpc, dscratch0, dscratch1, dddc, dpcc, dscratch0c, dscratch1c, what are the 5 I'm missing?

Comment (tariqkurd-repo) [reply src/debug-integration.adoc:145]: One identified problem here was if we introduced an otype later, or something else, for compartmentalization. Would we need to redefine this CSR? The most basic idea was to have this as a CSR where the tag is _always_ set and so the software can write whatever it likes and have it tagged. This idea then extended to different root, or anything else that we may need in the future.

The other reason not to add more CSRs is that there are only 16 available. There are four standard ones: 
image

and we've added an alias to three of them, so that's 7. 
We've also added `dddc` and `dinfc`, making 9 so we're quite close to filling the whole space, and then there's no provision for adding any more without some kind of banking scheme.

So I'd really like to get the "make any cap" functionality based on a single debug CSR if possible.

File: src/debug-integration.adoc [hunk 9]
```diff
@@ -11,6 +11,25 @@ When executing code due to an abstract command, the hart stays in debug mode
...
 
+==== CHERI Mode when in Debug Mode
+
+When entering debug mode, the CHERI mode is always the same, and may be
+switchable depending on which extensions have been implemented.
+
+If {cheri_mode_ext_name} is implemented:
+
+* the core enters Legacy Mode when entering debug mode,
+ and the mode can be switched using <<MODESW>>.
```
Comment (sorear) [inline src/debug-integration.adoc:22]: Why is this in this PR?  It seems unrelated to non-monotonic capability construction.

This will be nontrivial to implement for Rocket-type debuggers with a Debug ROM that contains instructions in a fixed ISA.

File: src/debug-integration.adoc (hunk 6 above)
Comment (arichardson) [reply src/debug-integration.adoc:142]: Fair enough, happy to keep it RW with all written bits ignored.

File: src/debug-integration.adoc (hunk 9 above)
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:22]: because otherwise there's no way of switching to capability mode in debug mode. Previously MODESW was potentially illegal in debug mode.

Comment (jrtc27) [reply src/debug-integration.adoc:22]: I continue be of the belief that, just as DXLEN is fixed at the maximum supported, so too should debug mode on a CHERI system always be capability mode rather than legacy mode

Comment (tariqkurd-repo) [reply src/debug-integration.adoc:22]: Is it not reasonable on a legacy system that the debugger starts in legacy mode and then switches into cap mode? For software compatibility?

Comment (jrtc27) [reply src/debug-integration.adoc:22]: Then you make legacy the primary mode, which the spec is //not// doing in other places.

Comment (jrtc27) [reply src/debug-integration.adoc:22]: (and everything that wants to use capability mode, e.g. to read a capability from a CSR, has to waste precious instruction buffer space switching to capability mode)

**Review by arichardson — APPROVED** [review]

**Review by sorear — APPROVED** [review]
I still think that a Rocket-style debug ROM implementation will require D mode to always execute using the capability ISA, but that can be discussed as a separate issue.

We're no longer violating the debug spec's own rules on PC access, so approve.

File: src/debug-integration.adoc [hunk 10]
```diff
@@ -1,17 +1,20 @@
...
+ for CHERI-RISC-V so change is likely.
+
+==== Debug Mode
 
 When executing code due to an abstract command, the hart stays in debug mode
 and the rules outlined in Section 4.1 of cite:[riscv-debug-spec]
 apply.
 
-=== Core Debug Registers
+==== Core Debug Registers
```
Comment (andresag01) [inline src/debug-integration.adoc:17]: I think the headers within this section are a little off.

**Review by andresag01 — COMMENTED** [review]
Just a small formatting change. Otherwise looks good to me.

File: src/debug-integration.adoc (hunk 10 above)
Comment (tariqkurd-repo) [reply src/debug-integration.adoc:17]: yeah - there aren't enough levels to make it work - I think it's good enough

File: src/debug-integration.adoc (hunk 9 above)
Comment (tariqkurd-repo) [reply src/debug-integration.adoc]: cap mode is the default now, and MODESW is optionally supported
